docs/enhancements/agent-upgrade.md

```diff
@@ -0,0 +1,256 @@
+---
+title: agent-upgrade
+authors:
+  - "@jhernand"
+creation-date: 2022-06-13
+last-updated: 2022-06-13
+---
+
+# Agent upgrade
+
+## Summary
+
+Usually the time since the agent is started and the cluster is installed is
+short. But in some situations it may be long, even many days, and during that
+long time the service may be upgraded and may require a newer version of the
+agent. There is currently no mechanism to upgrade the agent, the service
+silently works with the older agent assuming that nothing will go wrong. The
+only way to upgrade the agent would be to reboot the host, and we don't ask
+users to do that. Even if we did, it would be inconvenient, specially if there
+are many hosts or if rebooting a host requires manual intervention. This is
+particularly common in the late binding scenario. This is a proposal to
+implement a mechanism to upgrade the agent without having to reboot the host.
+
+Note that all new agent versions require a rollout of the service pods (as the
+agent image name is delivered as an environment variable to the service). Not
+all rollouts of service pods come with a newer agent version.
+
+## Motivation
+
+Reduce compatibility concerns when designing and developing new features.
+
+Reduce the need to reboot hosts when there are changes in the service that
+require new versions of the agent.
+
+### Goals
+
+The agent image will be automatically upgraded to the version expected by the
+service before proceeding to the installation of the cluster.
+
+Note that in this context _version_ is really a container image name, like
+`registry-proxy.engineering.redhat.com/rh-osbs/openshift4-assisted-installer-agent-rhel8:v1.0.0-141`.
+
+### Non-Goals
+
+The upgrade feature will not work for changes in the part of the agent that
+performs the initial registration since the upgrade part relies on the
+registration mechanism.
+
+Upgrading other images like the image of the service itself of the image of the
+Kubernetes controller is completely out of the scope of this enhancement.
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As a developer I want the agent to be automatically upgraded to what the service
+expects, so that I can assume that the service will never proceed to install a
+cluster with a host that is running an incompatible version of the agent.
+
+### Implementation Details/Notes/Constraints
+
+A new `compatible-agent` validation will be added to the service to check if the
+agent is compatible.
+
+The logic for that validation will be a comparison of the value of the
```

Comment:
It's important to note that the validation should automatically pass when the host is not in the relevant states (e.g. installing)

Reply:
That's a good point